collector/lib/Inotify.h

```diff
@@ -0,0 +1,292 @@
+   * use the WatcherNotFound method with the value returned from this
+   * method.
+   *
+   * @param needle The path we want the Watcher for.
+   * @returns An iterator to the Watcher.
+   */
+  WatcherIterator FindWatcher(const std::filesystem::path& needle);
+
+  /**
+   * Look for a Watcher for the given watch descriptor.
+   *
+   * Useful for searching for a Watcher from the inotify_event.wd
+   * field when processing an event.
+   *
+   * In order to check if the returned iterator is valid, callers must
+   * use the WatcherNotFound method with the value returned from this
+   * method.
+   *
+   * @param needle The path we want the Watcher for.
+   * @returns An iterator to the Watcher.
+   */
+  WatcherIterator FindWatcher(int needle);
+
+  /**
+   * Validate if the given iterator is in the Watcher store.
+   *
+   * @param w An iterator, usually gotten by calling FindWatcher.
+   * @returns true if the given iterator is not in the store.
+   */
+  bool WatcherNotFound(WatcherIterator w) { return w == watchers_.end(); }
+
+  /**
+   * Read events from the inotify subsystem.
+   *
+   * The events are read in bulk, but returned one at a time for the
+   * caller to process. Successive calls will either return another
+   * event that was read on a previous call or attempt to read more
+   * events. If no event is read after some time, InotifyTimeout will
+   * be returned instead, allowing the caller to do some additional
+   * processing before calling GetNext again.
+   *
+   * @returns One of three variants:
+   *          - inotify_event* when an event was successfully read.
+   *          - InotifyError when a system error prevented reading events.
+   *          - InotifyTimeout if everything is working as expected but
+   *            no events were read after some time.
+   */
+  InotifyResult GetNext();
+
+  /**
+   * Translate an inotify_event->mask to a printable string.
+   *
+   * @returns A string in the format '<event_type> | <other_type>...'
+   */
+  static std::string MaskToString(uint32_t mask);
+
+ private:
+  int fd_;
+  std::array<char, 1024> buffer_{};
+  size_t curr_byte_{0};
+  size_t read_data_{0};
+
+  // Watchers should probably be organized in a map. However,
+  // sometimes we'll want to look for these with a path, sometimes
+  // with a wd, so instead of over complicating things and assuming we
+  // will only have a couple watchers for now, we just do a vector and
+  // find by iterating. If we ever need more than just a couple, we
+  // MUST change this.
+  WatcherStore watchers_;
+
+  friend std::ostream& operator<<(std::ostream& os, const Inotify& inotify) {
+    bool first = true;
+    for (const auto& w : inotify.watchers_) {
+      if (first) {
+        first = false;
+      } else {
+        os << "\n";
+      }
+
+      os << w;
+    }
+    return os;
+  }
+
+  // Taken from <sys/inotify.h>
+  static constexpr std::array<unsigned int, 17> BIT_MASKS{
```

Comment:
Why copy from there?

Reply:
Because this way I can have an iterable array, rather than unrelated `#define` statements that mean nothing on their own.

Reply:
I thought this exact code was in `<sys/inotify.h>`. I checked `<sys/inotify.h> and that is not the case. I think a map should be used here, so that you don't get a mismatch between `BIT_MASK` and `MASK_STR`. I think it is easy to mess up the order here.

Reply:
I thought you had copied `BIT_MASK` from `<sys/inotify.h>`, but actually looking at the file, I can see that is not the case.

Reply:
I can now see how the comment may be misleading, I'll tweak it a bit to make it clearer we are creating the array from the defines in that header.